### PR TITLE
Kajam 11.2.0 plat 3434

### PR DESCRIPTION
**Pull request by kobimichaeli**

### CODE REVIEW

**Comment by MosheMaorKaltura** [comment]
Add release notes and we are done!

File: plugins/collaajUpdatePlugin/CollaajUpdatePlugin.php [hunk 1]
```diff
@@ -0,0 +1,21 @@
+<?php
+
+class CollaajUpdatePlugin extends KalturaPlugin implements IKalturaServices {
+
+    const PLUGIN_NAME = "CollaajUpdate";
```
Comment (tan-tan-kanarek) [inline plugins/collaajUpdatePlugin/CollaajUpdatePlugin.php]: Plugin name starts with non-capital letter.

Comment (tan-tan-kanarek) [inline plugins/collaajUpdatePlugin/CollaajUpdatePlugin.php]: Also, we agreed to call it captureSpace.
@gonenradai, please correct me if I'm wrong.

File: plugins/collaajUpdatePlugin/CollaajUpdatePlugin.php [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+    public static function getPluginName() {
+        return self::PLUGIN_NAME;
+    }
+
+    public static function getServicesMap()
+    {
+        $map = array(
+            'collaajini' => 'CollaajUpdateService',
```
Comment (tan-tan-kanarek) [inline plugins/collaajUpdatePlugin/CollaajUpdatePlugin.php]: We agreed also to call the service captureSpace.

File: plugins/collaajUpdatePlugin/services/CollaajUpdateService.php [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    }
+
+
+    /**
+     * Collaaj check for update
+     *
+     * @action serveUpdate
+     * @param string $os
+     * @param string $version
+     * @return KalturaCaptureSpaceUpdateResponse
```
Comment (tan-tan-kanarek) [inline plugins/collaajUpdatePlugin/services/CollaajUpdateService.php]: serve actions return file, not KalturaCaptureSpaceUpdateResponse.

File: plugins/collaajUpdatePlugin/lib/api/KalturaCaptureSpaceUpdateResponse.php [hunk 4]
```diff
@@ -0,0 +1,18 @@
+<?php
+
+class KalturaCaptureSpaceUpdateResponse extends KalturaObject {
```
Comment (tan-tan-kanarek) [inline plugins/collaajUpdatePlugin/lib/api/KalturaCaptureSpaceUpdateResponse.php]: What's the purpose of this class, it's just wrapping info.
Why won't it contain the content of info instead?

File: plugins/collaajUpdatePlugin/services/CollaajUpdateService.php [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    {
+        $collaajini = new collaajini();
+        $returned_object = new KalturaCaptureSpaceUpdateResponse();
+        $result = $collaajini->returnUpdateFileUrl($os, $version);
+
+        if ($result) {
+            $res = new KalturaCaptureSpaceUpdateResponse();
+            $info = new KalturaCaptureSpaceUpdateResponseInfo();
+            $info->url = $result['url'];
+            $info->hash = new KalturaCaptureSpaceUpdateResponseInfoHash();
```
Comment (tan-tan-kanarek) [inline plugins/collaajUpdatePlugin/services/CollaajUpdateService.php]: If hash contains a KalturaCaptureSpaceUpdateResponseInfoHash then change it's type definition, currently it's defined as string.

File: plugins/collaajUpdatePlugin/lib/api/KalturaCaptureSpaceUpdateResponse.php (hunk 4 above)
Comment (kobimichaeli) [inline plugins/collaajUpdatePlugin/lib/api/KalturaCaptureSpaceUpdateResponse.php]: This was not done by me, but rather Gonen.
I don’t know to which end.

Thanks,
Kobi M.

File: plugins/collaajUpdatePlugin/lib/api/KalturaCaptureSpaceUpdateResponseInfoHash.php [hunk 6]
```diff
@@ -0,0 +1,24 @@
+<?php
+
+class KalturaCaptureSpaceUpdateResponseInfoHash extends KalturaObject {
+    /**
+     * @var string
+     */
+    public $algorithm;
```
Comment (tan-tan-kanarek) [inline plugins/collaajUpdatePlugin/lib/api/KalturaCaptureSpaceUpdateResponseInfoHash.php]: Can you use KalturaAppTokenHashType instead of string?

Comment (tan-tan-kanarek) [inline plugins/collaajUpdatePlugin/lib/api/KalturaCaptureSpaceUpdateResponseInfoHash.php]: Or something similar?

File: plugins/collaajUpdatePlugin/lib/model/collaajInstallData.php [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+// This class handles all the data queries relevant for colaaj and the ini file
+class collaajInstallData
+{
+    private $ini_handle = "";  // handle to the ini object
+    private $ini_content = "";
+    private $ini_sorted_versions_array = array ();
+
+    public function __construct($ini_file_name){
+        if ($ini_file_name) {
+            $this->ini_handle = new iniReader($ini_file_name);
```
Comment (tan-tan-kanarek) [inline plugins/collaajUpdatePlugin/lib/model/collaajInstallData.php]: Why don't you use with simple old parse_ini_file?

File: plugins/collaajUpdatePlugin/lib/model/collaajInstallData.php [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+            $this->ini_content = $this->ini_handle->getIni_content();
+            $this->shift_array_lables_to_version($this->ini_handle->getIni_content());
+        } else throw new Exception ("ini file name was not provided");
+    }
+
+
+    public function getLatestVersionData ($needed_extension) {
```
Comment (tan-tan-kanarek) [inline plugins/collaajUpdatePlugin/lib/model/collaajInstallData.php]: This method shouldn't be public.

File: plugins/collaajUpdatePlugin/lib/model/collaajini.php [hunk 9]
```diff
@@ -0,0 +1,96 @@
+<?php
+
+class collaajini
```
Comment (tan-tan-kanarek) [inline plugins/collaajUpdatePlugin/lib/model/collaajini.php]: Why did you separate collaajini class from collaajInstallData class?

Comment (tan-tan-kanarek) [inline plugins/collaajUpdatePlugin/lib/model/collaajini.php]: This entire class looked to me complicated with no need, just hold the values in array and return them if they exists.
Also, I don't think it should be placed under model.

File: plugins/collaajUpdatePlugin/lib/model/collaajini.php [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+            }
+            $this->returned_data["url"] = $this->prepareDownloadUrl($given_os,$current_version, "serveupdate",self::DOWNLOAD_URL_TEMPLATE.$this->returned_data[self::DOWNLOAD_URL]);
+            return $this->returned_data;
+
+        } else throw new KalturaAPIException ("There is no update for $current_version");
+    }
+
+    public function prepareDownloadUrl($needed_os, $current_version, $action, $original_url) {
+        $returned_url = $original_url;
+        $returned_url = str_replace ("<HOSTNAME>", gethostname(),$returned_url);
```
Comment (tan-tan-kanarek) [inline plugins/collaajUpdatePlugin/lib/model/collaajini.php]: Hostname means nothing beyond firewall or load-balancer, use the common functionality to return URLs to API calls, see getURL of other serve actions.

File: plugins/collaajUpdatePlugin/services/CollaajUpdateService.php [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+            $info->hash->algorithm = "md5";
+            $info->hash->value = $result['md5'];
+            $res->info = $info;
+
+            //$returned_object->fromObject($collaajini);
+            //return $returned_object;
+            return $res;
+        } else throw new KalturaAPIException ("No update is available for version " . $version . " for " . $os);
```
Comment (tan-tan-kanarek) [inline plugins/collaajUpdatePlugin/services/CollaajUpdateService.php]: No need for else after return.

File: plugins/collaajUpdatePlugin/services/CollaajUpdateService.php [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+     * @return KalturaCaptureSpaceUpdateResponse
+     * @throws exception
+     */
+    public function serveInstallAction($os, $version)
+    {
+        /* @var $fileSync FileSync */
+        $collaajini = new collaajini();
+        $returned_object = new KalturaCaptureSpaceUpdateResponse();
+        $filePath = $collaajini->returnLatestVersionUrl($os, $version, "serveinstall");
+        $t = $collaajini->getVersion();
```
Comment (tan-tan-kanarek) [inline plugins/collaajUpdatePlugin/services/CollaajUpdateService.php]: 't' stands for what?

File: plugins/collaajUpdatePlugin/services/CollaajUpdateService.php [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+     * @action serveUpdate
+     * @param string $os
+     * @param string $version
+     * @return KalturaCaptureSpaceUpdateResponse
+     * @throws exception
+     */
+    public function serveUpdateAction($os, $version)
+    {
+
+        $collaajini = new collaajini();
```
Comment (tan-tan-kanarek) [inline plugins/collaajUpdatePlugin/services/CollaajUpdateService.php]: If all actions instantiate $collaajini, you could do it in initService.

File: plugins/collaajUpdatePlugin/lib/model/collaajini.php [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+            foreach ($ini_returned_data as $key => $value) {
+                $this->returned_data[$key] = $value;
+            }
+            // Appending the current file name to the to future url
+            $this->returned_data["url"] = $this->prepareDownloadUrl($given_os,$current_version, "serveinstall",self::DOWNLOAD_URL_TEMPLATE.$this->returned_data[self::DOWNLOAD_URL]);
+            return $this->returned_data[self::DOWNLOAD_URL];
+        } else throw new KalturaAPIException ("No install version found for $given_os");
+    }
+
+    public function returnUpdateFileUrl($given_os, $current_version) {
```
Comment (tan-tan-kanarek) [inline plugins/collaajUpdatePlugin/lib/model/collaajini.php]: Please rename to getUpdateFileUrl

File: plugins/collaajUpdatePlugin/lib/model/collaajini.php [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+        "download_url" => "",
+        "md5" => "",
+    );
+
+    public function __construct() {
+        $relative_ini_path = dirname(__FILE__) .'/'.self::INI_FILE_NAME;
+        $this->collaaj_data = new collaajInstallData($relative_ini_path);
+    }
+
+    public function returnLatestVersionUrl($given_os, $current_version) {
```
Comment (tan-tan-kanarek) [inline plugins/collaajUpdatePlugin/lib/model/collaajini.php]: Please rename to getLatestVersionUrl.

File: plugins/collaajUpdatePlugin/lib/model/collaajini.php [hunk 16]
```diff
@@ -0,0 +1,96 @@
+<?php
+
+class collaajini
+{
+    const INI_FILE_NAME = "../../collaajClientUpdate.ini";
+    const AVAILABLE_OS_TYPES = "Mac OS X,Windows";
+    const UPDATE_KEY = "version";
+    const DOWNLOAD_URL = "download_url";
+    const MD5 = "md5";
+    const DOWNLOAD_URL_TEMPLATE = "http://<HOSTNAME>/api_v3/index.php/service/collaajupdate_collaajupdate/action/<ACTION>/os/<OS>/version/<VERSION>/filename/";
```
Comment (tan-tan-kanarek) [inline plugins/collaajUpdatePlugin/lib/model/collaajini.php]: What if the action called in https, or with a different port?

File: plugins/collaajUpdatePlugin/CollaajUpdatePlugin.php (hunk 1 above)
Comment (gonenradai) [inline plugins/collaajUpdatePlugin/CollaajUpdatePlugin.php]: the feature in Kaltura is named "capturespace"
I don't really mind how the API itself is called, but for the sake of consistency - keeping the same name makes sense.

File: plugins/collaajUpdatePlugin/lib/api/KalturaCaptureSpaceUpdateResponse.php (hunk 4 above)
Comment (gonenradai) [inline plugins/collaajUpdatePlugin/lib/api/KalturaCaptureSpaceUpdateResponse.php]: @tan-tan-kanarek 
The XML we agreed on with Collaaj had nested nodes.
to make our API response output such an XML, we need an object within an object, and this is why we had to add a property which is of this "info" type.

File: plugins/collaajUpdatePlugin/lib/api/KalturaCaptureSpaceUpdateResponseInfoHash.php (hunk 6 above)
Comment (gonenradai) [inline plugins/collaajUpdatePlugin/lib/api/KalturaCaptureSpaceUpdateResponseInfoHash.php]: if the value of that enum is eventually a sting such as "md5" that the client can "read" and use to run the relevant algorithm to verify the file authenticity - sure, no problem from integration perspective.

**Pull request merged**